### ISSUE TITLE
networkd: add new Token= mode to use lower 64 bits of IPv6-LL address for global addr

### ISSUE THREAD

**Issue by perkelix** [systemd/systemd#33866]
### systemd version the issue has been seen with

256.4, 252.26

### Used distribution

Debian testing, stable

### Linux kernel version used

6.9.10, 6.1.99

### CPU architectures issue was seen on

None

### Component

systemd-networkd

### Expected behaviour you didn't see

1) I expected networkd to use the local-link IPv6 address generated by the kernel.
2) I expected networkd to notice that the local-link IPv6 address generated by the kernel has the stable-privacy flag and to use it to generate the public IPv6 address.

### Unexpected behaviour you saw

1) networkd generated a new local-link IPv6 address, in addition to the existing one.
2) networkd ignored both local-link addresses' stable-privacy flags and generated an EUI64 public address.

### Steps to reproduce the problem

Reproducible on both x86-32 and x86-64 hosts, using the systemd and Linux kernel versions as above.

1) Create the following /etc/systemd/network/dhcp.network unit:
```
[Match]
Name=en* wl*

[Network]
DHCP=yes
IPv6PrivacyExtensions=yes
IPv6LinkLocalAddressGenerationMode=stable-privacy
```
2) Enable networkd via systemctl and reboot the host.
3) '`ip a`' shows that the interface has two` fe80:: `addresses, both with the `stable-privacy` flag enabled, while the `2001::` address to the outside world doesn't have the `stable-privacy` flag and visibly was generated using `EUI64`.
4) Notice that none of the above should have happened given the following `sysctl.conf` fragment:
```
# Generate IPv6 Stable Privacy addesses instead of EUI64.
net.ipv6.conf.all.addr_gen_mode=3
net.ipv6.conf.default.addr_gen_mode=3
# Enable IPv6 Privacy. Prefer random over EUI64 derived.
net.ipv6.conf.all.use_tempaddr=2
net.ipv6.conf.default.use_tempaddr=2
```

**Comment by yuwata** [comment]
networkd does not generate any IPv6 link-local addresses, but the kernel generates it. If multiple IPv6 link-local addresses are assigned on an interface, then that should be a bug in the kernel. Please report to the kernel's bug tracker.

**Comment by perkelix** [comment]
networkd very much sets the flags it wants the kernel to use, and it does it incorrectly. Other clients don't have that issue.

**Comment by perkelix** [comment]
Here's a journalctl dump and 'ip a' on a host with a single network device.

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute 
       valid_lft forever preferred_lft forever
2: enp0s13: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN group default qlen 1000
    link/ether 00:40:ca:9d:f0:50 brd ff:ff:ff:ff:ff:ff
    inet 172.16.1.13/16 metric 1024 brd 172.16.255.255 scope global enp0s13
       valid_lft forever preferred_lft forever
    inet6 2001:14ba:a0fa:5405:e268:e935:df2a:789e/64 scope global temporary dynamic 
       valid_lft 2214sec preferred_lft 2214sec
    inet6 2001:14ba:a0fa:5405:240:caff:fe9d:f050/64 scope global dynamic mngtmpaddr noprefixroute 
       valid_lft 2214sec preferred_lft 2214sec
    inet6 fe80::fc53:80a4:5b52:b774/64 scope link stable-privacy 
       valid_lft forever preferred_lft forever
    inet6 fe80::dfab:fbc6:a59d:60c2/64 scope link stable-privacy 
       valid_lft forever preferred_lft forever
```

[networkd_log.txt](https://github.com/user-attachments/files/16422183/networkd_log.txt)

**Comment by yuwata** [comment]
> Expected behaviour you didn't see
> 2. I expected networkd to notice that the local-link IPv6 address generated by the kernel has the stable-privacy flag and to use it to generate the public IPv6 address.
> Unexpected behaviour you saw
> 2. networkd ignored both local-link addresses' stable-privacy flags and generated an EUI64 public address.

Note, though it is not equivalent what you expect, but you can use `Token=` setting in `[IPv6AcceptRA]` section to generate custom global address.

**Comment by perkelix** [comment]
Which really doesn't fix networkd's default behavior.

**Comment by yuwata** [comment]
We can add another `Token=` mode to use lower 64bit of IPv6-LL address for global address. But,
1. Why do you want that? Are there any RFC or so that suggest the way?
2. Even we implement that, the default should not be changed to keep backward compat.

**Comment by perkelix** [comment]
Please read again what the bug reported. I'm not asking for networkd to implement yet another token. I'm expecting networkd to use the privacy flags set in the unit file or, if none are set there, to respect the flags already configured for the interface. It doesn't do that. Other DHCP/RA clients have no problem doing this.

**Comment by yuwata** [comment]
Sorry, I cannot follow.

- `IPv6LinkLocalAddressGenerationMode=` controls `addr_gen_mode`. It _only_ affects for IPv6-LL address generated by the kernel, and is irrelevant for the global addresses.
- Global addresses are generated based on the `Token=` setting, and the default is EUI64.
- `IPv6PrivacyExtensions=` setting controls if a temporary address is generated for a global address.

About the issue that multiple IPv6-LL addresses are assigned for the interface, again, that's a bug in the kernel, as we neither generate nor assign them. You can see the log that these addresses are 'foreign', that means networkd does not control them.

For the report about the global address, I cannot find anything strange in the above result. An EUI64 is generated by networkd, and a temporary address is generated by the kernel. It looks expected.

**Comment by yuwata** [comment]
> Global addresses are supposed to derive from the LL one. If the LL address is EUI64-based, the global one is expected to use its suffix. If the LL address is stable-privacy-based, the global one is expected to use that suffix.

As I asked in the above, are there any RFC or so to request or suggest that?

**Comment by perkelix** [comment]
Actually, even setting `Token=prefixstable` in `[IPv6AcceptRA]` doesn't fix the problem. The global address still doesn't use stable-privacy:

`inet6 2001:14ba:a0fa:5405:94c7:4ef5:1f0:653f/64 scope global dynamic mngtmpaddr noprefixroute `

**Comment by perkelix** [comment]
Additionally, the only way to make networkd stop adding a second LL address is to change `IPv6LinkLocalAddressGenerationMode` from `stable-privacy` to `random`.

**Comment by yuwata** [comment]
> Actually, even setting `Token=prefixstable` in `[IPv6AcceptRA]` doesn't fix the problem. The global address still doesn't use stable-privacy:
> 
> `inet6 2001:14ba:a0fa:5405:94c7:4ef5:1f0:653f/64 scope global dynamic mngtmpaddr noprefixroute `

That's expected. `prefixstable` does not mean using lower 64 bits of IPv6-LL address.

**Comment by perkelix** [comment]
For comparison, here's what I get booting off an older systemd release (230) on an older Linux kernel (4.9.88) that doesn't implement stable-privacy and presumes EUI64 for everything:

```
2: enp0s13: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UNKNOWN group default qlen 1000
    link/ether 00:40:ca:9d:f0:50 brd ff:ff:ff:ff:ff:ff
    inet 172.16.1.13/16 brd 172.16.255.255 scope global enp0s13
       valid_lft forever preferred_lft forever
    inet6 2001:14ba:a0fa:5405:5dbc:8be4:7036:685c/64 scope global temporary dynamic 
       valid_lft 568sec preferred_lft 567sec
    inet6 2001:14ba:a0fa:5405:240:caff:fe9d:f050/64 scope global mngtmpaddr dynamic 
       valid_lft 568sec preferred_lft 567sec
    inet6 fe80::240:caff:fe9d:f050/64 scope link 
       valid_lft forever preferred_lft forever
```

This output remains consistent across the board. The suffix is the same for the link and global addresses. This is what's expected.

**Comment by yuwata** [comment]
By default, IPv6-LL is generated by the kernel is EUI64. So, there is nothing strange.

**Comment by perkelix** [comment]
My comment wasn't only about LL. It was about consistent address generation for the same interface.

**Comment by yuwata** [comment]
Both IPv6-LL and global addresses are generated with EUI64 by default. IIRC, that's not changed for long time. I cannot find anything strange.

Sorry, but still I cannot follow what you expects.

**Comment by perkelix** [comment]
I expect networkd to consistently use the same address generation method for all interfaces that match, for both LL and global e.g. If stable-privacy is configured, use that for both LL and global; if random is configured, use that for both LL and global.

**Comment by yuwata** [comment]
That's not currently supported. That's all. Not a bug.

**Comment by perkelix** [comment]
It very much is a bug.  It deters from what other network configuration tools e.g. dhcpcd, dhclient, ifupdown have been doing for a much longer time than systemd has existed. Those tools have a very consistent way of configuring LL and global addresses for an interface. networkd doesn't.

**Comment by bluca** [comment]
The fact that legacy cruft like ifupdown or dhcpcd do something doesn't mean it's a bug not doing the exact same. Given Yu already asked for RFCs indicating this is a requirement and nothing has been provided, it seems clear this is not a bug, but an RFE based on some personal preferences regarding how things should work. It's fine to file RFEs to add new options and modes of course.

**Comment by perkelix** [comment]
I'd like to see RFCs that demonstrate that networkd's way of configuring LL and global inconsistantly is RFC compliant. That would add credibility to claims of other network configuration tools being mere "legacy cruft."

**Comment by bluca** [comment]
I don't think demanding that a negative be proven is very productive. As already mentioned, it's fine to ask for new features/options/modes.

**Comment by perkelix** [comment]
I don't think that presuming that systemd is RFC-compliant until proven otherwise and calling everyone else's code "legacy cruft" is productive at all.

**Comment by perkelix** [comment]
I also don't think that changing the bug's title to match what you THINK I meant is productive either.

The issues, in order:

1. `net.ipv6.conf.*.addr_gen_mode=3` and `net.ipv6.conf.*.use_tempaddr=2 `are configured by sysctl during boot PER INTERFACE, but networkd ignores those settings and instead inconsistently adds addresses with or without these flags.
2. networkd adds a superfluous LL address if anything else than `random` is configured as `IPv6LinkLocalAddressGenerationMode`.
3. The address generation mode between LL and global addresses is inconsistent. This should not be the case, since the kernel configures this PER INTERFACE.

**Comment by bluca** [comment]
The ticket is not titled after what you thought or meant, nor it's supposed to be. It's titled after what is reasonable to ask, which is what Yu already outlined in https://github.com/systemd/systemd/issues/33866#issuecomment-2257560267

**Comment by perkelix** [comment]
It doesn't respect the kernel settings. THIS IS A BUG. Please stop renaming other people's bug reports. THANK YOU.

**Comment by bluca** [comment]
> It doesn't respect the kernel settings.

As already explained, the kernel settings are for link local addresses, not global addresses. I understand that does not meet your expectation, but absent any RFC mandating the contrary, it's just that, an expectation, and it doesn't mean there's any bug if your personal preferences aren't met. Again, a new optional token mode to support your use case is a perfectly reasonable RFE, and if anybody sends a PR to implement that, we'll happily review it and merge it.

> Please stop renaming other people's bug reports

Please understand that this is not _your_ bug report, it is _our_ bug tracker. We as maintainers name, rename, split, merge, open, close, label, relabel tickets as it best helps us do our jobs.

**Comment by perkelix** [comment]
I have very much said that the issue affects both LL and global addresses. networkd systematically fails at creating global addresses that use stable-privacy. It also does numerous fishy things with LL. Other clients don't have that problem.

**Comment by bluca** [comment]
As it has been explained already, link locals are created by the kernel, not networkd. And global addresses are created with stable-privacy - which doesn't mean using the same bits as link locals. As explained in the documentation, by default EUI64 is used, and prefixstable can be specified, which absent any more specific configuration a hash keyed on the machine-id and interface name will be used, so that the result _is_ stable _and_ private:

https://www.freedesktop.org/software/systemd/man/latest/systemd.network.html#prefixstable%5B:ADDRESS%5D%5B,UUID%5D

You imply that stable necessarily has to mean the same bits from the LL address has to be used, but again, such claims need evidence, and legacy tools doing that by chance isn't evidence. You can point at any time to an RFC supporting your preference, and then this would become a bug. Until that is the case, it remains an RFE for a new option.

**Comment by perkelix** [comment]
Please stop bringing this back to the sole issue of LL.  A number of networkd issues related to its IPv6 implementation were mentioned in this bug report. LL is just one of them.

Personally, I don't care how networkd needs to be configured to accomplish a result similar to what other clients do. I'm perfectly fine with being pointed to the right configuration options to end up with stable-privacy for both LL and Global, and calling it a day. 

Neither you or Yu have given me that. Instead, both of you keep on moving the goalpost and shifitng the focus of my bug report to something else than what I brought up. Please stop doing that.

**Comment by yuwata** [comment]
> Please stop bringing this back to the sole issue of LL. A number of networkd issues related to its IPv6 implementation were mentioned in this bug report. LL is just one of them.

As explained, the issue for LL is a bug in the kernel, the other one is expected behavior unless it is RFC incompliant.

> Personally, I don't care how networkd needs to be configured to accomplish a result similar to what other clients do. I'm perfectly fine with being pointed to the right configuration options to end up with stable-privacy for both LL and Global, and calling it a day.

As already mentioned multiple times, `Token=prefixstable` can be used for global address, and `IPv6LinkLocalAddressGenerationMode=stable-privacy` for LL.

> Neither you or Yu have given me that. Instead, both of you keep on moving the goalpost and shifitng the focus of my bug report to something else than what I brought up. Please stop doing that.

Only what we do not support currently is that the lower 64 bits of LL and global addresses are not identical.
Do you find any issue caused by the mismatch of the lower 64 bits of the addresses? You have not explained what kind of 'bug' is caused by the mismatch.

**Comment by perkelix** [comment]
As already explained many times:

1. `IPv6LinkLocalAddressGenerationMode=stable-privacy` ADDS a new LL address, rather than use the existing LL address that already has the stable-privacy flag. The only generation mode that doesn't ADD a superfluous LL address is `random`. You cannot call this a kernel bug since other network configuration tools don't produce that behaviour.
2. `Token=prefixstable` doesn't enable the `stable-privacy `flag for the global address.

**Comment by yuwata** [comment]
> As already explained many times:
> 
>     1. `IPv6LinkLocalAddressGenerationMode=stable-privacy` ADDS a new LL address, rather than use the existing LL address that already has the stable-privacy flag. The only generation mode that doesn't ADD a superfluous LL address is `random`. You cannot call this a kernel bug since other network configuration tools don't produce that behaviour.

Already explained many times that networkd does not generate any IPv6-LL addresses. Other tools are other tools. That cannot be a evidence that this is a bug in networkd.

>     2. `Token=prefixstable` doesn't enable the `stable-privacy` flag for the global address.

The flag is assigned by the kernel. Cannot set from the userspace. And, the flag means the address is generated by the **kernel's** stable-privacy logic. Even if a userspace tool assigns an address generated with a stable-privacy logic, the flag is not set for the address.


Please read all above replies and understand what the kernel does and networkd does. And the two points you said is on the kernel side. There is nothing we can do.

**Comment by bluca** [comment]
networkd does _not_ use the kernel to generate the global addresses on RA, intentionally - as we need to get some information out of the RA packets. So it cannot set the flags that the kernel owns. But as already explained, and well documented, this does not mean that the generated addresses are not compliant with the stable and private RFC - in fact, they are perfectly compatible with RFC 7217 which, once again, does _not_ mandate or even suggest that the same bits as the stable-private link local addresses on the same interface have to be used. That's an implementation details of some other tools, the absence of which is _not_ a bug, just a missing feature - that is fine to add, once again.

**Comment by arvidjaar** [comment]
@yuwata 

> * `IPv6LinkLocalAddressGenerationMode=` controls `addr_gen_mode`. It _only_ affects for IPv6-LL address generated by the kernel, and is irrelevant for the global addresses.

It is not what the kernel documentation says.

> addr_gen_mode - INTEGER
Defines how link-local **and autoconf** addresses are generated.

So, it is pretty much relevant for the global addresses derived from RA.

Maybe you mean, that 

> networkd does _not_ use the kernel to generate the global addresses on RA

In which case the least surprise behavior would still be to default to the current `addr_gen_mode` on the interface unless explicitly overridden. This would result in the same behavior whether using in-kernel autoconfig or networkd autoconfig. And that really is what this issue is about.

I also do not understand how the "use lower 64 bits of IPv6-LL address for global addr" ever came in picture at all. The original complaint was

> 2\. I expected networkd to notice that the local-link IPv6 address generated by the kernel has the stable-privacy flag and to use it to generate the public IPv6 address.

I see the confusion what "it" above refers to. I understood it was referring to the "stable-privacy flag" and not to the "link-local IPv6 address". Which effectively is equivalent to "respect the current `addr_gen_mode` on the interface". Maybe @perkelix can clarify.

**Comment by bluca** [comment]
> It is not what the kernel documentation says.

Yes, it is. The kernel talks about its own network system, so it applies _if you are using the kernel to generate slaac addresses_ - which is not the case here. I can use iproute2 to manually set addresses if I like, and they would not respect that statement either - it doesn't mean iproute2 is "broken".

> In which case the least surprise behavior would still be to default to the current `addr_gen_mode` on the interface unless explicitly overridden. This would result in the same behavior whether using in-kernel autoconfig or networkd autoconfig.

No, that would definitely not be "least surprise", in fact it would be "most surprise", given the kernel setting is obviously for the kernel, and then this is not a new project, it already exists, so changing behaviour like that would be extremely surprising. Again, networkd does not use the kernel to generate slaac addresses, as already explained, and it has its own extensive configuration system. One might not like it, but it is most definitely not a "bug" nor "broken".

> I see the confusion what "it" above refers to. I understood it was referring to the "stable-privacy flag" and not to the "link-local IPv6 address". Which effectively is equivalent to "respect the current addr_gen_mode on the interface".

No, this was actually explained quite clearly, and it does not refer merely to the flag, but to using the same suffix across local and global addresses:

> Global addresses are supposed to derive from the LL one. If the LL address is EUI64-based, the global one is expected to use its suffix. If the LL address is stable-privacy-based, the global one is expected to use that suffix. Other DHCP/RA clients indeed do so. networkd doesn't.

https://github.com/systemd/systemd/issues/33866#issuecomment-2257596805

which again is not mandated or even suggested by any RFC, but it's simply an implementation detail of the kernel, when it generates both addresses. And again, it's fine to have as an RFE for a new mode, but it definitely does not mean "networkd is broken" as it was originally claimed.

**Comment by perkelix** [comment]
@arvidjaar understood my bug report correctly. The changed title indeed no longer reflacts that.
